### ISSUE TITLE
How to optionally match?

### ISSUE THREAD

**Issue by bluenote10** [alehander92/gara#22]
Coming from Scala's pattern matching, I'm a bit confused what the library is doing. In Scala, pattern matching performs validation, and whether something matches is optional. Picking up on the email example, how can I actually validate whether the input type satisfies the criteria for unwrapping the values?

```nim
match("not a valid email string"):
  email(name, domain):
    echo &"Valid email: {name}@{domain}"
  _:
    echo "not a valid email"
```

Specifically, how should I express that `s.split('@', 1)` may or may not work? Is it only possible to unwrap things unconditionally with this library or is there a mechanism for expression optional matches?

My expectation from Scala were to write an `unapply` function (what is called unpackers here), which can return either
- a bool for direct value matching
- an `Option[T]` for unwrapping a single value, or
- an `Option[(T, ...)]` for unwrapping multiple values.
For matching an email for instance, the `unapply` implementation would get an input string, and return `some((name, domain))` only if `s.split('@', 1)` is valid.

**Comment by alehander92** [comment]
i agree this might make more sense, i think i intended to use e.g. empty seq / invalid values for cases when i want to make sure the unpacker doesnt match, but it's not a good solution indeed.

i still find Option[<T: having length/index>] simpler than having both Option[T] and Option[(T, ..)] but i suspect the single value case might be a sensible optimisation..
